File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <string>
#include <stack>
using namespace std;

int main(void)
{
	int t;
	stack<char> s;
	string input;
	string answer = "YES\n";

	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	cin >> t;
	for (int i = 0; i < t; i++)
	{
		answer = "YES\n";
		cin >> input;
		for (unsigned int j = 0; j < input.size(); j++)
		{
			if (input[j] == '(')
				s.push('(');
			else
			{
				if (s.empty())
				{
					answer = "NO\n";
					break ;
				}
				s.pop();
			}
		}
		if (!s.empty())
			answer = "NO\n";
		for (;!s.empty();)
			s.pop();
		cout << answer;
	}
	return (0);
}

File: 백준/Silver/9012. 괄호/괄호.py
test_case = int(input())

for i in range(test_case):
    next_flag = False
    input_string = input()
    input_list = []
    for char in input_string:
        if char == '(':
            input_list.append('(')
        else:
            if len(input_list) == 0:
                print("NO")
                next_flag = True
                break
            input_list.pop()
    if next_flag:
        continue
    if len(input_list) != 0:
        print("NO")
    else:
        print("YES")
